### PR TITLE
Linear implementation for Fulkerson-Chen-Anstee

### LINKED ISSUE #1452: is_graphical: linear-time test for simple directed case
The Fulkerson-Ryser test, used in `is_graphical` for simple directed graphs, can (and should) be checked in linear time (assuming that the degree sequence was already sorted). Currently, we have a quadratic time implementation.

See section 5.1 in:

H. Kim, C. I. Del Genio, K. E. Bassler, and Z. Toroczkai, New J. Phys. 14, 023012 (2012). https://dx.doi.org/10.1088/1367-2630/14/2/023012

### PR DESCRIPTION
**Pull request by Tagl**
Closes #1452

The current implementation of the Fulkerson-Chen-Anstee simple directed graphicality check is quadratic.
This is my own approach to the linear time version of Fulkerson-Chen-Anstee.
I (with @szhorvat) plan to do a write-up for this approach.

After implementing the method I read section 5.1 in:
H. Kim, C. I. Del Genio, K. E. Bassler, and Z. Toroczkai, New J. Phys. 14, 023012 (2012). https://dx.doi.org/10.1088/1367-2630/14/2/023012
The methods follow the same general approach.
The key difference is they use recurrence relations to prove that the method works.
I leave here an explanation of how my method works and as in the other methods, the first step is sorting.

The theorem (screenshotted from Wikipedia):

image

The left hand side is easy to compute in linear time for each value of k.

On the right hand side we have two sums of outdegrees.
For the first sum we separate outdegrees based on whether they are $< k-1$ or $>= k-1$.
Similarly, for the second sum we separate outdegrees based on whether they are $< k$ or $>= k$.
In both cases we have a value which we can think of as the splitting point of a sum.
We only need to track the sum of outdegrees below the splitting point, because the value of k is incrementing from $1$ to $n$, and therefore they will not individually be impacted by future changes.

The outdegrees that are equal to or above the splitting point must be tracked, and we do so with an ordered multiset.
The ordered set for the first sum is initially empty and the one for the second sum initially contains all of the outdegrees.
As $k$ increases, either add/remove the outdegree at index k-1 to/from the sum or ordered set, depending on which side of the splitting point it lies.
We may also need to remove the smallest outdegrees from the ordered sets and move them over to the sums of outdegrees below the splitting point.
The contribution of the values in the ordered multiset to the right hand side is the value of $k-1$ or $k$, depending on which sum we are considering, multiplied by the size of the ordered multiset.

Using a balanced binary search tree for the ordered set means adding and removing from the set takes $\mathcal{O}(\log n)$ time.
Then the time complexity will match that of the sort operation.

We can push this a little further.
Since the values in the set are bounded by n, we can switch the quicksort out for a bucket sort which is $\mathcal{O}(n)$.
Additionally, it allows us to switch out a balanced binary search tree for an array of integers, where the index in to the array represents the degree and the value at that index represents how many times that degree appears in the ordered multiset.
All operations within the main loop are $\mathcal{O}(1)$ and therefore the time complexity of the algorithm is $\mathcal{O}(n)$ in total.
I have not done any benchmarking yet, but I have tested an equivalent C++ implementation on sequences of size $1$ to $2\,000\,000$ and there is a noticeable difference in running time between the linear and the linearithmic approaches.

### CODE REVIEW

File: src/misc/graphicality.c [hunk 1]
```diff
@@ -627,17 +603,35 @@ static igraph_error_t igraph_i_is_graphical_directed_simple(const igraph_vector_
...
-    sort_data.second = out_degrees;
+    for (i = 0; i < vcount; i++) {
+        if (VECTOR(*in_degrees)[i] >= vcount || VECTOR(*out_degrees)[i] >= vcount) {
+            *res = false;
+            igraph_vector_int_list_destroy(&buckets);
+            IGRAPH_FINALLY_CLEAN(1);
+            return IGRAPH_SUCCESS;
+        }
+        current_bucket = igraph_vector_int_list_get_ptr(&buckets, VECTOR(*in_degrees)[i]);
+        igraph_vector_int_push_back(current_bucket, VECTOR(*out_degrees)[i]);
```
Comment (ntamas) [inline src/misc/graphicality.c:618]: We are probably missing an `IGRAPH_CHECK` here (unless you can ensure that the operation never needs to reallocate the underlying memory, in which case this should be documented).

**Review by ntamas — COMMENTED** [review]
A few comments, nothing conclusive yet. I've talked to @szhorvat and he will also want to chime in and review later.

File: src/misc/graphicality.c (hunk 1 above)
Comment (Tagl) [reply src/misc/graphicality.c:618]: That is correct, will fix

**Comment by szhorvat** [comment]
I edited the theorem into the top post so it's easier to follow your description, and what the LHS and RHS refer to.

File: src/misc/graphicality.c [hunk 2]
```diff
@@ -627,17 +603,35 @@ static igraph_error_t igraph_i_is_graphical_directed_simple(const igraph_vector_
...
 
-    /* Set up the auxiliary struct for sorting */
-    sort_data.first  = in_degrees;
-    sort_data.second = out_degrees;
+    for (i = 0; i < vcount; i++) {
+        if (VECTOR(*in_degrees)[i] >= vcount || VECTOR(*out_degrees)[i] >= vcount) {
+            *res = false;
+            igraph_vector_int_list_destroy(&buckets);
+            IGRAPH_FINALLY_CLEAN(1);
+            return IGRAPH_SUCCESS;
```
[the remark this answers is not in the document]
Comment (Tagl) [reply src/misc/graphicality.c]: I have applied the changes you suggested.

**Comment by szhorvat** [comment]
I added some rudimentary benchmarks.

----

Before this PR:

```
|> Benchmark file: /Users/szhorvat/Repos/igraph-main/igraph/tests/benchmarks/graphicality.c
| Undirected simple,       PA, n = 50, 2000000x                                     0.27s   0.27s      0s
| Undirected loops,        PA, n = 50, 2000000x                                    0.635s  0.633s      0s
| Undirected multi,        PA, n = 50, 2000000x                                    0.055s  0.055s      0s
| Undirected multi, loops, PA, n = 50, 2000000x                                        0s      0s      0s
| Undirected simple,       G(n,p), n = 50, 2000000x                                  0.2s  0.198s      0s
| Undirected loops,        G(n,p), n = 50, 2000000x                                0.789s  0.786s      0s
| Undirected multi,        G(n,p), n = 50, 2000000x                                0.053s  0.053s      0s
| Undirected multi, loops, G(n,p), n = 50, 2000000x                                    0s      0s      0s
| Directed simple,       PA, n = 50, 2000000x                                       1.26s   1.26s      0s
| Directed loops,        PA, n = 50, 2000000x                                       1.18s   1.17s      0s
| Directed multi,        PA, n = 50, 2000000x                                      0.061s   0.06s      0s
| Directed multi, loops, PA, n = 50, 2000000x                                          0s      0s      0s
| Directed simple,       G(n,p), n = 50, 2000000x                                   1.24s   1.23s      0s
| Directed loops,        G(n,p), n = 50, 2000000x                                   1.17s   1.17s      0s
| Directed multi,        G(n,p), n = 50, 2000000x                                  0.061s   0.06s      0s
| Directed multi, loops, G(n,p), n = 50, 2000000x                                      0s      0s      0s

| Undirected simple,       PA, n = 1000, 100000x                                    0.13s  0.128s      0s
| Undirected loops,        PA, n = 1000, 100000x                                   0.544s   0.54s      0s
| Undirected multi,        PA, n = 1000, 100000x                                   0.059s  0.058s      0s
| Undirected multi, loops, PA, n = 1000, 100000x                                       0s      0s      0s
| Undirected simple,       G(n,p), n = 1000, 100000x                               0.101s    0.1s      0s
| Undirected loops,        G(n,p), n = 1000, 100000x                               0.659s  0.654s      0s
| Undirected multi,        G(n,p), n = 1000, 100000x                               0.059s  0.058s      0s
| Undirected multi, loops, G(n,p), n = 1000, 100000x                                   0s      0s      0s
| Directed simple,       PA, n = 1000, 100000x                                      1.83s   1.82s      0s
| Directed loops,        PA, n = 1000, 100000x                                     0.825s  0.821s      0s
| Directed multi,        PA, n = 1000, 100000x                                      0.06s  0.059s      0s
| Directed multi, loops, PA, n = 1000, 100000x                                         0s      0s      0s
| Directed simple,       G(n,p), n = 1000, 100000x                                   1.8s    1.8s      0s
| Directed loops,        G(n,p), n = 1000, 100000x                                 0.827s  0.825s      0s
| Directed multi,        G(n,p), n = 1000, 100000x                                  0.06s  0.059s      0s
| Directed multi, loops, G(n,p), n = 1000, 100000x                                     0s      0s      0s

| Undirected simple,       PA, n = 1000000, 100x                                   0.152s  0.134s  0.017s
| Undirected loops,        PA, n = 1000000, 100x                                   0.995s  0.968s  0.021s
| Undirected multi,        PA, n = 1000000, 100x                                   0.059s  0.058s      0s
| Undirected multi, loops, PA, n = 1000000, 100x                                   0.058s  0.057s      0s
| Undirected simple,       G(n,p), n = 1000000, 100x                                0.12s  0.102s  0.017s
| Undirected loops,        G(n,p), n = 1000000, 100x                                1.42s   1.39s  0.022s
| Undirected multi,        G(n,p), n = 1000000, 100x                               0.059s  0.058s      0s
| Undirected multi, loops, G(n,p), n = 1000000, 100x                               0.058s  0.057s      0s
| Directed simple,       PA, n = 1000000, 100x                                      19.5s   19.4s  0.028s
| Directed loops,        PA, n = 1000000, 100x                                      1.47s   1.42s  0.042s
| Directed multi,        PA, n = 1000000, 100x                                     0.062s  0.061s      0s
| Directed multi, loops, PA, n = 1000000, 100x                                     0.059s  0.058s      0s
| Directed simple,       G(n,p), n = 1000000, 100x                                  19.6s   19.6s  0.028s
| Directed loops,        G(n,p), n = 1000000, 100x                                  1.47s   1.42s  0.043s
| Directed multi,        G(n,p), n = 1000000, 100x                                 0.061s   0.06s      0s
| Directed multi, loops, G(n,p), n = 1000000, 100x                                 0.059s  0.058s      0s
```

----

After this PR:

```
|> Benchmark file: /Users/szhorvat/Repos/igraph-main/igraph/tests/benchmarks/graphicality.c
| Undirected simple,       PA, n = 50, 2000000x                                    0.251s   0.25s      0s
| Undirected loops,        PA, n = 50, 2000000x                                      0.6s  0.598s      0s
| Undirected multi,        PA, n = 50, 2000000x                                    0.053s  0.053s      0s
| Undirected multi, loops, PA, n = 50, 2000000x                                        0s      0s      0s
| Undirected simple,       G(n,p), n = 50, 2000000x                                0.185s  0.183s      0s
| Undirected loops,        G(n,p), n = 50, 2000000x                                0.767s  0.764s      0s
| Undirected multi,        G(n,p), n = 50, 2000000x                                0.054s  0.053s      0s
| Undirected multi, loops, G(n,p), n = 50, 2000000x                                    0s      0s      0s
| Directed simple,       PA, n = 50, 2000000x                                       3.06s   3.05s      0s
| Directed loops,        PA, n = 50, 2000000x                                       1.13s   1.13s      0s
| Directed multi,        PA, n = 50, 2000000x                                      0.061s   0.06s      0s
| Directed multi, loops, PA, n = 50, 2000000x                                          0s      0s      0s
| Directed simple,       G(n,p), n = 50, 2000000x                                   3.06s   3.05s      0s
| Directed loops,        G(n,p), n = 50, 2000000x                                   1.14s   1.13s      0s
| Directed multi,        G(n,p), n = 50, 2000000x                                   0.06s   0.06s      0s
| Directed multi, loops, G(n,p), n = 50, 2000000x                                      0s      0s      0s

| Undirected simple,       PA, n = 1000, 100000x                                    0.13s  0.128s      0s
| Undirected loops,        PA, n = 1000, 100000x                                   0.542s  0.538s      0s
| Undirected multi,        PA, n = 1000, 100000x                                   0.059s  0.058s      0s
| Undirected multi, loops, PA, n = 1000, 100000x                                       0s      0s      0s
| Undirected simple,       G(n,p), n = 1000, 100000x                               0.101s    0.1s      0s
| Undirected loops,        G(n,p), n = 1000, 100000x                               0.656s  0.654s      0s
| Undirected multi,        G(n,p), n = 1000, 100000x                               0.059s  0.058s      0s
| Undirected multi, loops, G(n,p), n = 1000, 100000x                                   0s      0s      0s
| Directed simple,       PA, n = 1000, 100000x                                      2.23s   2.23s      0s
| Directed loops,        PA, n = 1000, 100000x                                     0.803s  0.799s      0s
| Directed multi,        PA, n = 1000, 100000x                                     0.061s  0.059s      0s
| Directed multi, loops, PA, n = 1000, 100000x                                         0s      0s      0s
| Directed simple,       G(n,p), n = 1000, 100000x                                  2.24s   2.23s      0s
| Directed loops,        G(n,p), n = 1000, 100000x                                 0.804s  0.801s      0s
| Directed multi,        G(n,p), n = 1000, 100000x                                  0.06s  0.059s      0s
| Directed multi, loops, G(n,p), n = 1000, 100000x                                     0s      0s      0s

| Undirected simple,       PA, n = 1000000, 100x                                   0.151s  0.133s  0.017s
| Undirected loops,        PA, n = 1000000, 100x                                   0.991s  0.966s  0.021s
| Undirected multi,        PA, n = 1000000, 100x                                   0.059s  0.058s      0s
| Undirected multi, loops, PA, n = 1000000, 100x                                   0.058s  0.057s      0s
| Undirected simple,       G(n,p), n = 1000000, 100x                                0.12s  0.103s  0.017s
| Undirected loops,        G(n,p), n = 1000000, 100x                                1.42s   1.39s  0.022s
| Undirected multi,        G(n,p), n = 1000000, 100x                               0.059s  0.058s      0s
| Undirected multi, loops, G(n,p), n = 1000000, 100x                               0.059s  0.057s      0s
| Directed simple,       PA, n = 1000000, 100x                                      2.54s   2.16s  0.357s
| Directed loops,        PA, n = 1000000, 100x                                      1.45s   1.39s  0.048s
| Directed multi,        PA, n = 1000000, 100x                                      0.06s   0.06s      0s
| Directed multi, loops, PA, n = 1000000, 100x                                     0.059s  0.058s      0s
| Directed simple,       G(n,p), n = 1000000, 100x                                  2.56s   2.16s   0.37s
| Directed loops,        G(n,p), n = 1000000, 100x                                  1.46s   1.39s  0.049s
| Directed multi,        G(n,p), n = 1000000, 100x                                 0.061s  0.061s      0s
| Directed multi, loops, G(n,p), n = 1000000, 100x                                  0.06s  0.059s      0s
```

**Comment by szhorvat** [comment]
It seems that at about 1000 vertices the old implementation is still doing better despite the bad complexity. This might be because there's a lot of memory allocation when using a list of vectors, and memory allocation is expensive. Although the system time (3rd column) doesn't increase significantly. Let's discuss next week what might be going on.

**Comment by szhorvat** [comment]
@Tagl I refactored the counting sort to use a single array instead of an array of arrays. Can you please have a look to make sure I didn't break anything?  Efficiency is greatly improved now:

```
|> Benchmark file: /Users/szhorvat/Repos/igraph-main/igraph/tests/benchmarks/graphicality.c
| Undirected simple,       PA,     n = 50, 2000000x                                 0.24s  0.239s      0s
| Undirected loops,        PA,     n = 50, 2000000x                                0.597s  0.594s      0s
| Undirected multi,        PA,     n = 50, 2000000x                                 0.05s  0.049s      0s
| Undirected multi, loops, PA,     n = 50, 2000000x                                    0s      0s      0s
| Undirected simple,       G(n,p), n = 50, 2000000x                                0.178s  0.175s      0s
| Undirected loops,        G(n,p), n = 50, 2000000x                                0.775s  0.771s      0s
| Undirected multi,        G(n,p), n = 50, 2000000x                                 0.05s  0.049s      0s
| Undirected multi, loops, G(n,p), n = 50, 2000000x                                    0s      0s      0s
| Directed simple,       PA,     n = 50, 2000000x                                   1.26s   1.26s      0s
| Directed loops,        PA,     n = 50, 2000000x                                   1.14s   1.14s      0s
| Directed multi,        PA,     n = 50, 2000000x                                  0.062s  0.061s      0s
| Directed multi, loops, PA,     n = 50, 2000000x                                      0s      0s      0s
| Directed simple,       G(n,p), n = 50, 2000000x                                   1.11s    1.1s      0s
| Directed loops,        G(n,p), n = 50, 2000000x                                    1.5s    1.5s      0s
| Directed multi,        G(n,p), n = 50, 2000000x                                  0.062s  0.061s      0s
| Directed multi, loops, G(n,p), n = 50, 2000000x                                      0s      0s      0s

| Undirected simple,       PA,     n = 1000, 100000x                               0.128s  0.126s      0s
| Undirected loops,        PA,     n = 1000, 100000x                               0.532s  0.529s      0s
| Undirected multi,        PA,     n = 1000, 100000x                               0.058s  0.057s      0s
| Undirected multi, loops, PA,     n = 1000, 100000x                                   0s      0s      0s
| Undirected simple,       G(n,p), n = 1000, 100000x                               0.101s    0.1s      0s
| Undirected loops,        G(n,p), n = 1000, 100000x                               0.647s  0.645s      0s
| Undirected multi,        G(n,p), n = 1000, 100000x                               0.058s  0.057s      0s
| Undirected multi, loops, G(n,p), n = 1000, 100000x                                   0s      0s      0s
| Directed simple,       PA,     n = 1000, 100000x                                 0.812s  0.809s      0s
| Directed loops,        PA,     n = 1000, 100000x                                 0.816s  0.814s      0s
| Directed multi,        PA,     n = 1000, 100000x                                  0.06s  0.059s      0s
| Directed multi, loops, PA,     n = 1000, 100000x                                     0s      0s      0s
| Directed simple,       G(n,p), n = 1000, 100000x                                 0.583s  0.581s      0s
| Directed loops,        G(n,p), n = 1000, 100000x                                  1.28s   1.28s      0s
| Directed multi,        G(n,p), n = 1000, 100000x                                  0.06s   0.06s      0s
| Directed multi, loops, G(n,p), n = 1000, 100000x                                     0s      0s      0s

| Undirected simple,       PA,     n = 1000000, 100x                               0.152s  0.133s  0.018s
| Undirected loops,        PA,     n = 1000000, 100x                                1.02s  0.992s  0.021s
| Undirected multi,        PA,     n = 1000000, 100x                               0.058s  0.058s      0s
| Undirected multi, loops, PA,     n = 1000000, 100x                               0.059s  0.058s      0s
| Undirected simple,       G(n,p), n = 1000000, 100x                               0.121s  0.102s  0.018s
| Undirected loops,        G(n,p), n = 1000000, 100x                                1.32s   1.29s  0.023s
| Undirected multi,        G(n,p), n = 1000000, 100x                               0.058s  0.058s      0s
| Undirected multi, loops, G(n,p), n = 1000000, 100x                               0.058s  0.057s      0s
| Directed simple,       PA,     n = 1000000, 100x                                     1s  0.852s  0.138s
| Directed loops,        PA,     n = 1000000, 100x                                  1.41s   1.36s  0.044s
| Directed multi,        PA,     n = 1000000, 100x                                  0.06s   0.06s      0s
| Directed multi, loops, PA,     n = 1000000, 100x                                 0.058s  0.058s      0s
| Directed simple,       G(n,p), n = 1000000, 100x                                 0.791s  0.627s  0.163s
| Directed loops,        G(n,p), n = 1000000, 100x                                  2.84s   2.79s  0.045s
| Directed multi,        G(n,p), n = 1000000, 100x                                  0.06s   0.06s      0s
| Directed multi, loops, G(n,p), n = 1000000, 100x                                 0.059s  0.058s      0s
```

**Comment by szhorvat** [comment]
@ntamas I think we're almost good to go with this. Any other comments? I ended up not using the single exit point you suggested because it makes it difficult to de-allocate memory early when we don't need it.

**Comment by szhorvat** [comment]
In principle a further optimization is possible since we already compute the cumulative in-degree sums, which are the LHS, during the counting sort, except that this code does it in the wrong order (from smallest to largest instead of largest to smallest, which the `lhs` is supposed to be). 

This should not block merging though, as it's not even clear if it improves performance. For the final version and explanation of the algorithm we should sort it out though.

**Comment by Tagl** [comment]
> I refactored the counting sort to use a single array instead of an array of arrays. Can you please have a look to make sure I didn't break anything?

Looks good to me. The benchmark numbers are also much more in line with the measurements of my C++ implementation.

**Review by ntamas — APPROVED** [review]

**Pull request merged**